test/common/common.go

```diff
@@ -0,0 +1,446 @@
+	t.Helper()
+
+	clientPod := buildDigPod("digpod", namespace, hostname)
+	if err := cl.Create(ctx, clientPod); err != nil {
+		return nil, fmt.Errorf("failed to create pod %s/%s: %v", clientPod.Namespace, clientPod.Name, err)
+	}
+	defer func() {
+		_ = cl.Delete(ctx, clientPod)
+	}()
+
+	// Loop until dig pod starts, then parse logs for query results.
+	var responseCode string
+	var gotIPs map[string]struct{}
+	err := wait.PollUntilContextTimeout(ctx, 5*time.Second, 10*time.Minute, true, func(ctx context.Context) (bool, error) {
+		if err := cl.Get(ctx, types.NamespacedName{Name: clientPod.Name, Namespace: clientPod.Namespace}, clientPod); err != nil {
+			t.Errorf("Failed to get pod %s/%s: %v", clientPod.Namespace, clientPod.Name, err)
+			return false, nil
+		}
+		switch clientPod.Status.Phase {
+		case corev1.PodRunning:
+			t.Log("Waiting for dig pod to finish")
+			return false, nil
+		case corev1.PodPending:
+			t.Log("Waiting for dig pod to start")
+			return false, nil
+		case corev1.PodFailed, corev1.PodSucceeded:
+			// Failed or Succeeded, let's continue on to check the logs.
+			break
+		default:
+			return true, fmt.Errorf("unhandled pod status type")
+		}
+
+		// Get logs of the dig pod.
+		readCloser, err := cs.CoreV1().Pods(clientPod.Namespace).GetLogs(clientPod.Name, &corev1.PodLogOptions{
+			Container: clientPod.Spec.Containers[0].Name,
+			Follow:    false,
+		}).Stream(ctx)
+		if err != nil {
+			t.Logf("Failed to read output from pod %s: %v (retrying)", clientPod.Name, err)
+			return false, nil
+		}
+		scanner := bufio.NewScanner(readCloser)
+		defer func() {
+			if err := readCloser.Close(); err != nil {
+				t.Fatalf("Failed to close reader for pod %s: %v", clientPod.Name, err)
+			}
+		}()
+
+		gotIPs = make(map[string]struct{})
+		for scanner.Scan() {
+			line := scanner.Text()
+
+			// Skip blank lines.
+			if strings.TrimSpace(line) == "" {
+				continue
+			}
+			// Parse status out (helpful for future debugging)
+			if strings.HasPrefix(line, ";;") && strings.Contains(line, "status:") {
+				responseCodeSection := strings.TrimSpace(strings.Split(line, ",")[1])
+				responseCode = strings.Split(responseCodeSection, " ")[1]
+				t.Logf("DNS Response Code: %s", responseCode)
+			}
+			// If it doesn't begin with ";", then we have an answer.
+			if !strings.HasPrefix(line, ";") {
+				splitAnswer := strings.Fields(line)
+				if len(splitAnswer) < 5 {
+					t.Logf("Expected dig answer to have 5 fields: %q", line)
+					return true, nil
+				}
+				gotIP := strings.Fields(line)[4]
+				gotIPs[gotIP] = struct{}{}
+			}
+		}
+		t.Logf("Got IPs: %v", gotIPs)
+		return true, nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to observe the expected dig results: %v", err)
+	}
+
+	// Need to delete pod in case we start again.
+	waitForDeletion(ctx, t, cl, clientPod, 5*time.Minute)
```

Comment:
Why do we need a differed delete of the POD set up before? Btw don't we want these logics to be the same?

Reply:
Ah yea good catch. I think this was here before I broke this into a new function call. Now the `defer` executes when the function exits.